app/react/Uploads/actions/uploadsActions.js

```diff
@@ -93,42 +93,42 @@ export function upload(docId, file, endpoint = 'upload') {
   });
 }
 
-export function publicSubmit(data) {
-  return dispatch => new Promise((resolve, reject) => {
-    const request = superagent.post(`${APIURL}public`)
+export function publicSubmit(data, remote = false) {
+  return dispatch => new Promise((resolve) => {
+    const request = superagent.post(remote ? `${APIURL}remotepublic` : `${APIURL}public`)
     .set('Accept', 'application/json')
     .set('X-Requested-With', 'XMLHttpRequest')
     .field('captcha', data.captcha);
-    delete data.captcha;
 
     if (data.file) {
       request.attach('file', data.file);
-      delete data.file;
     }
 
     if (data.attachments) {
       data.attachments.forEach((attachment, index) => {
         request.attach(`attachments[${index}]`, attachment);
-        delete data.attachments;
       });
     }
-
-    request.field('entity', JSON.stringify(data));
-
+    request.field('entity', JSON.stringify(Object.assign({}, { title: data.title, template: data.template, metadata: data.metadata })));
+    let completionResolve;
+    let completionReject;
+    const uploadCompletePromise = new Promise((_resolve, _reject) => { completionResolve = _resolve; completionReject = _reject; });
```

Comment:
The way I understand it, the main Promise returned by the action is resolved by the `progress` event of the request. So another promise is created to be resolved/rejected when the request is complete. And since the promise can't be returned directly from the event handler, the resolve/reject functions are store in the variables in order to be called later when the request is complete.

Reply:
As a follow-up, I found out the reason the initial promise is resolved on `progress` is so that the loading screen can be displayed during the upload.